### PR TITLE
[clang backend] In AArch64's DataLayout, specify a minimum function alignment of 4.

### PR DESCRIPTION
**Pull request by dougsonos**
This addresses an issue where the explicit alignment of 2 (for C++ ABI reasons) was being propagated to the back end and causing under-aligned functions (in special sections). (#90358)

This is an alternate approach suggested by @efriedma-quic in PR #90415.

~(There are a few failing tests. One I looked at was checking for the explicit alignment of 2, which no longer appears in AArch64 due to this change.)~

### CODE REVIEW

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 1]
```diff
@@ -5387,6 +5387,14 @@ std::string llvm::UpgradeDataLayoutString(StringRef DL, StringRef TT) {
     return Res;
   }
 
+  // AArch64 data layout upgrades.
+  if (T.isAArch64()) {
+    // Add "-Fn32"
+    if (!DL.contains("-Fn32"))
+      Res.append("-Fn32");
```
Comment (efriedma-quic) [inline llvm/lib/IR/AutoUpgrade.cpp:5394]: Ideally, the "fixed" datalayout string should be in the same order as the one a new compiler would generate (so if you're sticking it at the end here, stick it at the end on the actual strings).  I think we compare datalayouts using string equality in LTO.

**Review by efriedma-quic — COMMENTED** [review]
This looks correct.

I'd expect regression tests for the autoupgrade, and a test that clang isn't generating the alignment markings anymore.

File: llvm/lib/IR/AutoUpgrade.cpp (hunk 1 above)
Comment (dougsonos) [reply llvm/lib/IR/AutoUpgrade.cpp:5394]: OK, thanks. I was afraid of that. I was copying two patterns - in ARM, `Fn8` seemed to be after any `p` and before `i`. In this file, the modifications are usually (always?) appending.

Comment (efriedma-quic) [reply llvm/lib/IR/AutoUpgrade.cpp:5394]: Feel free to make the datalayout string whatever order is most convenient for the autoupgrade.

**Comment by dougsonos** [comment]
> I'd expect regression tests for the autoupgrade, and a test that clang isn't generating the alignment markings anymore.

I just pushed:
- the `-Fn32` is at the end of the data layout strings
- the 6 tests that failed for me locally (testing only AArch64) are fixed

That includes a change to CodeGen/CodeGenCXX/member-alignment.cpp, which checks for alignment / no alignment on C++ methods, depending on target.

I'm not clear on when the autoupgrade runs and how to test it. Is there an example I can follow?

Thanks!

**Comment by dougsonos** [comment]
Linux x86_64 has one test failure: RISCV/unnamed-sym-no-entry.c . AFAICT it looks unrelated.

File: llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp [hunk 2]
```diff
@@ -86,7 +86,7 @@ TEST(DataLayoutUpgradeTest, NoDataLayoutUpgrade) {
       "-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64"
       "-f80:128:128-n8:16:32:64-S128");
   EXPECT_EQ(DL2, "e-m:e-i64:64-n32:64");
-  EXPECT_EQ(DL3, "e-m:o-i64:64-i128:128-n32:64-S128");
+  EXPECT_EQ(DL3, "e-m:o-i64:64-i128:128-n32:64-S128-Fn32");
```
Comment (dougsonos) [inline llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp:89]: Actually this is supposed to be a no-op test; this should be moved and replaced with one which tests a no-op situation.

File: clang/lib/Basic/Targets/AArch64.cpp [hunk 3]
```diff
@@ -1480,11 +1480,11 @@ AArch64leTargetInfo::AArch64leTargetInfo(const llvm::Triple &Triple,
 void AArch64leTargetInfo::setDataLayout() {
   if (getTriple().isOSBinFormatMachO()) {
     if(getTriple().isArch32Bit())
-      resetDataLayout("e-m:o-p:32:32-i64:64-i128:128-n32:64-S128", "_");
+      resetDataLayout("e-m:o-p:32:32-i64:64-i128:128-n32:64-S128-Fn32", "_");
```
Comment (fhahn) [inline clang/lib/Basic/Targets/AArch64.cpp:1483]: Just to double check @TNorthover, is an alignment of 32 OK for arm64_32?

Comment (TNorthover) [reply clang/lib/Basic/Targets/AArch64.cpp:1483]: Yep. CPU would hard-fault if that was violated. All instructions are 4 bytes and must be aligned.

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by efriedma-quic** [comment]
Which email address do you want the commit associated with?  The GitHub "squash and merge" button wants to attribute it to your "sonosphere.com" address, but it looks like the commits themselves are using an "apple.com" address.

File: clang/lib/Basic/Targets/AArch64.cpp (hunk 3 above)
Comment (AtariDreams) [reply clang/lib/Basic/Targets/AArch64.cpp:1483]: Is this true even if the cpu was executing thumb2 instructions in AArch32 mode?

Comment (efriedma-quic) [reply clang/lib/Basic/Targets/AArch64.cpp:1483]: Thumb2 function pointers aren't aligned... but from LLVM's perspective it's a completely different target, so it's not relevant.

**Pull request merged**